source/tutorials/nixos-configuration-testing.md

```diff
@@ -0,0 +1,485 @@
+
+```nix
+machine.wait_for_unit("default.target")
+machine.succeed("su -- alice -c 'which firefox'")
+machine.fail("su -- root -c 'which firefox'")
+```
+
+
+The python script is referring to the `machine` attribute we used to name the configuration in the `nodes` attribute.
+The script waits until start up is reaching systemd `default.target`.
+We are using the `su` command to switch between users and the `which` command to see if the user has access to `firefox`.
+We expect that the command `which firefox` to succeed for user `alice` and to fail for `root`.
+This script is put into a function inside the `testScript` attribute.
+Unlike the virtual machine before we don't need to specify a password, the test script has super user rights in the virtual machine.
+
+The complete `minimaltest.nix` file content looks like the following:
+```{code-block}
+let
+  nixpkgs = <nixpkgs>;
+  pkgs = import nixpkgs {};
+in
+  pkgs.nixosTest {
+    nodes.machine = { config, pkgs, ... }: {
+      boot.loader.systemd-boot.enable = true;
+      boot.loader.efi.canTouchEfiVariables = true;
+
+      services.xserver.enable = true;
+      services.xserver.displayManager.gdm.enable = true;
+      services.xserver.desktopManager.gnome.enable = true;
+
+      users.users.alice = {
+        isNormalUser = true;
+        extraGroups = [ "wheel" ];
+        packages = with pkgs; [
+          firefox
+          thunderbird
+        ];
+      };
+
+      system.stateVersion = "22.05";
+    };
+    testScript = {nodes, ...}: ''
+      machine.wait_for_unit("default.target")
+      machine.succeed("su -- alice -c 'which firefox'")
+      machine.fail("su -- root -c 'which firefox'")
+    '';
+  }
+```
+
+### Running tests
+
+To set up all machines and execute the test script:[^nixdev]
+
+```shell-session
+$ nix-build minimaltest.nix
+```
+
+You'll notice an error message if something goes wrong.[^nixdev]
+
+In case the tests succeed, you should see at the end:[^nixdev]
+
+```shell-session
+...
+test script finished in 10.96s
+cleaning up
+killing machine (pid 10)
+(0.00 seconds)
+/nix/store/bx7z3imvxxpwkkza10vb23czhw7873w2-vm-test-run-unnamed
+```
+
+### interactive python shell to interact with virtual machine
+
+When developing tests or when something breaks, it’s useful to interactively fiddle with the script or access a terminal for a machine.[^nixdev]
+
+To interactively start a Python session with a testing framework:[^nixdev]
+```shell-session
+$ $(nix-build -A driverInteractive minimaltest.nix)/bin/nixos-test-driver
+```
+
+You can run any of the testing operations. The testScript attribute from our minimaltest.nix definition can be executed with `test_script()` function.[^nixdev]
+
+Within this Python shell you can enter a interactive shell you can run python commands like those in the test script.
+
+On the first evaluation of the command the virtual machine is started.
+You can manually trigger the start of the virtual machine by using
```

Comment:
```suggestion
Start a virtual machine with
```

[Use imperative for instructions.](https://github.com/NixOS/nix.dev/blob/d32652b94fcc1ed4336f3a8ebf7fc5678b48c1eb/CONTRIBUTING.md?plain=1#L8)